File: examples/inference_pipeline_repo/seed_repo/service.js
import { predict } from "./src/pipeline.js";

const protocol = "shinka-candidate-v1";
let pending = Buffer.alloc(0);

function frame(value) {
  const payload = Buffer.from(JSON.stringify(value), "utf8");
  const header = Buffer.alloc(4);
  header.writeUInt32BE(payload.length, 0);
  return Buffer.concat([header, payload]);
}

function send(value) {
  process.stdout.write(frame(value));
}

function consume() {
  while (pending.length >= 4) {
    const size = pending.readUInt32BE(0);
    if (size === 0 || size > 64 * 1024 * 1024) process.exit(2);
    if (pending.length < size + 4) return;
    const value = JSON.parse(pending.subarray(4, size + 4).toString("utf8"));
    pending = pending.subarray(size + 4);
    if (value.type !== "request" || typeof value.id !== "string") process.exit(2);
    try {
      send({ type: "response", id: value.id, output: predict(value.input) });
    } catch (error) {
      send({ type: "error", id: value.id, error: "candidate request failed" });
    }
  }
}

process.stdin.on("data", (chunk) => {
  pending = Buffer.concat([pending, chunk]);
  consume();
});
send({ type: "ready", protocol });

File: examples/circle_packing_repo_experiment_n41/seed_repo/service.py
"""Minimal framed-stdio service for the secure candidate runtime."""

from __future__ import annotations

import json
import struct
import sys

import numpy as np

from src.packing import run_packing


MAX_FRAME_BYTES = 64 * 1024 * 1024


def _send(value: dict) -> None:
    payload = json.dumps(value, separators=(",", ":"), allow_nan=False).encode()
    if not payload or len(payload) > MAX_FRAME_BYTES:
        raise RuntimeError("response frame is too large")
    sys.stdout.buffer.write(struct.pack(">I", len(payload)))
    sys.stdout.buffer.write(payload)
    sys.stdout.buffer.flush()


def _read_frame() -> dict | None:
    header = sys.stdin.buffer.read(4)
    if not header:
        return None
    if len(header) != 4:
        raise RuntimeError("truncated request header")
    size = struct.unpack(">I", header)[0]
    if size == 0 or size > MAX_FRAME_BYTES:
        raise RuntimeError("invalid request frame size")
    payload = sys.stdin.buffer.read(size)
    if len(payload) != size:
        raise RuntimeError("truncated request payload")
    value = json.loads(payload)
    if not isinstance(value, dict):
        raise RuntimeError("request must be an object")
    return value


def _serialise_result() -> dict:
    centers, radii, reported_sum = run_packing()
    centers = np.asarray(centers, dtype=float)
    radii = np.asarray(radii, dtype=float)
    return {
        "centers": centers.tolist(),
        "radii": radii.tolist(),
        "reported_sum": float(reported_sum),
    }


def main() -> None:
    _send({"type": "ready", "protocol": "shinka-candidate-v1"})
    while True:
        request = _read_frame()
        if request is None:
            return
        if request.get("type") != "request" or not isinstance(request.get("id"), str):
            raise RuntimeError("invalid request")
        try:
            _send({"type": "response", "id": request["id"], "output": _serialise_result()})
        except Exception:
            _send({"type": "error", "id": request["id"], "error": "candidate failed"})


if __name__ == "__main__":
    main()
